File: week-01/mario-more/mario.c
#include <stdio.h>

#define PYRAMID_HEIGHT_MIN 1
#define PYRAMID_HEIGHT_MAX 8
#define PYRAMID_GAP 2
#define PYRAMID_BLOCK '#'
#define AIR_BLOCK ' '

#include "../../lib/common.h"

void write_pyramid(int height);
void write_pyramid_gap();
int get_pyramid_height(int min, int max);

int main() {
	int height = get_pyramid_height(PYRAMID_HEIGHT_MIN, PYRAMID_HEIGHT_MAX);
	write_pyramid(height);
}

int get_pyramid_height(int min, int max) {
	int height = 0;

	do {
		printf("Height: ");
		height = get_int_input();
	} while (height < min || height > max);

	return height;
}

void write_pyramid_gap() {
	for (int gap = PYRAMID_GAP; gap > 0; gap--) {
		putchar(AIR_BLOCK);
	}
}

void write_pyramid(int height) {
	//      #  #
	//     ##  ##
	//    ###  ###
	//   ####  ####
	//  #####  #####
	// ######  ######

	for (int row = 1; row <= height; row++) {
		int air1 = height - row;
		int block1 = height - air1;
		int block2 = block1;

		while (air1) {
			putchar(AIR_BLOCK);
			air1--;
		}
		while (block1) {
			putchar(PYRAMID_BLOCK);
			block1--;
		}

		write_pyramid_gap();

		while (block2) {
			putchar(PYRAMID_BLOCK);
			block2--;
		}

		putchar('\n');
	}

	fflush(stdout);
}

File: week-06/sentimental-mario-more/mario.py
import sys
from typing import Final

PYRAMID_HEIGHT_MIN: Final = 1
PYRAMID_HEIGHT_MAX: Final = 8
PYRAMID_GAP: Final = 2
PYRAMID_BLOCK: Final = '#'
AIR_BLOCK: Final = ' '

def main():
	height = get_pyramid_height(PYRAMID_HEIGHT_MIN, PYRAMID_HEIGHT_MAX)
	write_pyramid(height)

def get_pyramid_height(min: int, max: int) -> int:
	while True:
		height = input_int("Height: ")
		if not (height < min or height > max):
			break

	return height


def write_pyramid_gap():
	for _ in range(PYRAMID_GAP):
		sys.stdout.write(AIR_BLOCK)


def write_pyramid(height: int):
	#      #  #
	#     ##  ##
	#    ###  ###
	#   ####  ####
	#  #####  #####
	# ######  ######

	for row in range(1, height + 1):
		air1 = height - row
		block1 = height - air1
		block2 = block1

		while (air1):
			sys.stdout.write(AIR_BLOCK)
			air1 -= 1
		while (block1):
			sys.stdout.write(PYRAMID_BLOCK)
			block1 -= 1
		
		write_pyramid_gap()

		while (block2):
			sys.stdout.write(PYRAMID_BLOCK)
			block2 -= 1

		sys.stdout.write('\n')


def input_int(prompt: str = '', err_msg: str = '') -> int:
	while True:
		try:
			return int(input(prompt))
		except ValueError:
			if err_msg != '':
				print(err_msg)


if __name__ == "__main__":
	main()

